File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/core/src/Public/Extensions/Storage/StorageProto+Helpers.ts
/**
 * StorageProto+Helpers.ts
 *
 * Ergonomic helpers for canonical Storage proto types.
 *
 * Mirrors Swift `StorageProto+Helpers.swift`. Pure-rename field aliases were
 * already removed on the Swift side (SWIFT-DUP-STORAGE-ALIASES) and are not
 * ported; only the helpers with actual logic are mirrored.
 */

import {
  DeviceStorageInfo as DeviceStorageInfoMessage,
  AppStorageInfo as AppStorageInfoMessage,
  StorageInfo as StorageInfoMessage,
  ModelStorageMetrics as ModelStorageMetricsMessage,
  StorageAvailability as StorageAvailabilityMessage,
} from '@runanywhere/proto-ts/storage_types';
import type {
  DeviceStorageInfo,
  AppStorageInfo,
  StorageInfo,
  ModelStorageMetrics,
  StorageAvailability,
} from '@runanywhere/proto-ts/storage_types';

/**
 * Build a `DeviceStorageInfo` with `usedPercent` derived from the byte
 * counts. Mirrors Swift `RADeviceStorageInfo.init(totalBytes:freeBytes:usedBytes:)`.
 */
export function makeDeviceStorageInfo(
  totalBytes: number,
  freeBytes: number,
  usedBytes: number
): DeviceStorageInfo {
  return DeviceStorageInfoMessage.fromPartial({
    totalBytes,
    freeBytes,
    usedBytes,
    usedPercent: totalBytes > 0 ? (usedBytes / totalBytes) * 100.0 : 0.0,
  });
}

/**
 * Build an `AppStorageInfo`. Mirrors Swift
 * `RAAppStorageInfo.init(documentsBytes:cacheBytes:appSupportBytes:totalBytes:)`.
 */
export function makeAppStorageInfo(
  documentsBytes: number,
  cacheBytes: number,
  appSupportBytes: number,
  totalBytes: number
): AppStorageInfo {
  return AppStorageInfoMessage.fromPartial({
    documentsBytes,
    cacheBytes,
    appSupportBytes,
    totalBytes,
  });
}

/** Empty storage snapshot. Mirrors Swift `RAStorageInfo.empty`. */
export function emptyStorageInfo(): StorageInfo {
  return StorageInfoMessage.fromPartial({
    app: AppStorageInfoMessage.fromPartial({}),
    device: DeviceStorageInfoMessage.fromPartial({}),
    models: [],
    totalModels: 0,
    totalModelsBytes: 0,
  });
}

/**
 * Sum of per-model on-disk sizes. Mirrors Swift
 * `RAStorageInfo.totalModelsSizeBytes`.
 */
export function totalModelsSizeBytes(info: StorageInfo): number {
  return info.models.reduce(
    (total, metrics) => total + metrics.sizeOnDiskBytes,
    0
  );
}

/**
 * Total models size with the per-model sum as fallback. Mirrors Swift
 * `RAStorageInfo.totalModelsSize`.
 */
export function totalModelsSize(info: StorageInfo): number {
  return info.totalModelsBytes > 0
    ? info.totalModelsBytes
    : totalModelsSizeBytes(info);
}

/**
 * Usage percentage of a device storage snapshot. Mirrors Swift
 * `RADeviceStorageInfo.usagePercentage`.
 */
export function usagePercentage(device: DeviceStorageInfo): number {
  if (device.totalBytes <= 0) return 0;
  return (device.usedBytes / device.totalBytes) * 100.0;
}

/**
 * Build a `ModelStorageMetrics`. Mirrors Swift
 * `RAModelStorageMetrics.init(modelID:sizeOnDiskBytes:lastUsedMs:)`.
 */
export function makeModelStorageMetrics(
  modelId: string,
  sizeOnDiskBytes: number,
  lastUsedMs?: number
): ModelStorageMetrics {
  return ModelStorageMetricsMessage.fromPartial({
    modelId,
    sizeOnDiskBytes,
    ...(lastUsedMs !== undefined ? { lastUsedMs } : {}),
  });
}

/**
 * Build a `StorageAvailability`. Mirrors Swift
 * `RAStorageAvailability.make(isAvailable:requiredBytes:availableBytes:recommendation:)`.
 */
export function makeStorageAvailability(
  isAvailable: boolean,
  requiredBytes: number,
  availableBytes: number,
  recommendation?: string
): StorageAvailability {
  return StorageAvailabilityMessage.fromPartial({
    isAvailable,
    requiredBytes,
    availableBytes,
    ...(recommendation !== undefined ? { recommendation } : {}),
  });
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/Storage/StorageProto+Helpers.swift
//
//  StorageProto+Helpers.swift
//  RunAnywhere SDK
//
//  Ergonomic helpers for canonical Storage proto types.
//

// MARK: - RADeviceStorageInfo

extension RADeviceStorageInfo {
    public init(totalBytes: Int64, freeBytes: Int64, usedBytes: Int64) {
        self.init()
        self.totalBytes = totalBytes
        self.freeBytes = freeBytes
        self.usedBytes = usedBytes
        self.usedPercent = totalBytes > 0
            ? Float(Double(usedBytes) / Double(totalBytes) * 100.0)
            : 0.0
    }

    // Aliases `totalSpace` / `freeSpace` / `usedSpace` removed — shadowed
    // the canonical proto field names (`totalBytes`/`freeBytes`/`usedBytes`)
    // without semantic value. Per swift.md SWIFT-DUP-STORAGE-ALIASES.

    public var usagePercentage: Double {
        guard totalBytes > 0 else { return 0 }
        return Double(usedBytes) / Double(totalBytes) * 100.0
    }
}

// MARK: - RAAppStorageInfo

extension RAAppStorageInfo {
    public init(documentsBytes: Int64, cacheBytes: Int64, appSupportBytes: Int64, totalBytes: Int64) {
        self.init()
        self.documentsBytes = documentsBytes
        self.cacheBytes = cacheBytes
        self.appSupportBytes = appSupportBytes
        self.totalBytes = totalBytes
    }

    // `documentsSize` / `cacheSize` / `appSupportSize` / `totalSize` aliases
    // removed. Per swift.md SWIFT-DUP-STORAGE-ALIASES — canonical proto
    // field names (`documentsBytes`/`cacheBytes`/`appSupportBytes`/
    // `totalBytes`) are what app code should use.
}

// MARK: - RAStorageInfo

extension RAStorageInfo {
    public static let empty: RAStorageInfo = {
        var info = RAStorageInfo()
        info.app = RAAppStorageInfo()
        info.device = RADeviceStorageInfo()
        info.models = []
        info.totalModels = 0
        info.totalModelsBytes = 0
        return info
    }()

    public var totalModelsSizeBytes: Int64 {
        models.reduce(0) { $0 + $1.sizeOnDiskBytes }
    }

    public var appStorage: RAAppStorageInfo {
        get { app }
        set { app = newValue }
    }

    public var deviceStorage: RADeviceStorageInfo {
        get { device }
        set { device = newValue }
    }

    public var totalModelsSize: Int64 {
        totalModelsBytes > 0 ? totalModelsBytes : totalModelsSizeBytes
    }

    public var modelCount: Int { models.count }

}

// MARK: - RAModelStorageMetrics

extension RAModelStorageMetrics {
    public init(modelID: String, sizeOnDiskBytes: Int64, lastUsedMs: Int64? = nil) {
        self.init()
        self.modelID = modelID
        self.sizeOnDiskBytes = sizeOnDiskBytes
        if let lastUsedMs { self.lastUsedMs = lastUsedMs }
    }

    // `modelId` / `sizeOnDisk` / `lastUsed` aliases removed — pure renames of
    // canonical proto field names (`modelID`/`sizeOnDiskBytes`/`lastUsedMs`)
    // with no consumer usage. Per swift.md SWIFT-DUP-STORAGE-ALIASES.
}

// MARK: - RAStorageAvailability

extension RAStorageAvailability {
    // `hasWarning` / `requiredSpace` / `availableSpace` aliases removed — pure
    // renames of canonical proto field names (`hasWarningMessage`/
    // `requiredBytes`/`availableBytes`). Per swift.md SWIFT-DUP-STORAGE-ALIASES.

    public static func make(
        isAvailable: Bool,
        requiredBytes: Int64,
        availableBytes: Int64,
        recommendation: String? = nil
    ) -> RAStorageAvailability {
        var availability = RAStorageAvailability()
        availability.isAvailable = isAvailable
        availability.requiredBytes = requiredBytes
        availability.availableBytes = availableBytes
        if let rec = recommendation { availability.recommendation = rec }
        return availability
    }
}
